ordinary_data/valve/tr_valve_pipe.sql

```diff
@@ -67,23 +114,35 @@ COMMENT ON TRIGGER tr_valve_pipe_delete ON qwat_od.pipe IS 'Trigger: when deleti
 
 
 /* AFTER UPDATE OF A VALVE: REEVALUATE OLD AND NEW PIPES */
-CREATE OR REPLACE FUNCTION qwat_od.ft_valve_update() RETURNS TRIGGER AS
+CREATE FUNCTION qwat_od.ft_valve() RETURNS TRIGGER AS
 $BODY$
 	BEGIN
-		PERFORM qwat_od.fn_pipe_update_valve(OLD.id);
-		PERFORM qwat_od.fn_pipe_update_valve(NEW.id);
+		IF TG_OP = 'INSERT' THEN
+			PERFORM qwat_od.fn_pipe_update_valve(NEW.id);
+		END IF;
+		IF TG_OP = 'DELETE' THEN
+			PERFORM qwat_od.fn_pipe_delete_valve(OLD.id);
+		END IF;
+		IF TG_OP = 'UPDATE' THEN
+			PERFORM qwat_od.fn_pipe_delete_valve(OLD.id);
+			PERFORM qwat_od.fn_pipe_update_valve(NEW.id);
```

Comment:
yep, if you can give details why we need two steps here.

Reply:
For one case: when you change the fk_pipe id for a valve